WordPress/MediaEditor/MediaEditor.swift

```diff
@@ -21,30 +27,59 @@ public class MediaEditor: UINavigationController {
     public var styles: MediaEditorStyles = [:] {
         didSet {
             currentCapability?.apply(styles: styles)
+            hub.apply(styles: styles)
         }
     }
 
     init(_ image: UIImage) {
         self.image = image
-        hub = MediaEditorHub()
         super.init(rootViewController: hub)
-        presentIfSingleImageAndCapability()
-        modalTransitionStyle = .crossDissolve
-        modalPresentationStyle = .fullScreen
-        navigationBar.isHidden = true
+        setup()
+    }
+
+    init(_ asyncImage: AsyncImage) {
+        self.asyncImage = asyncImage
+        super.init(rootViewController: hub)
+        setup()
     }
 
     required init?(coder aDecoder: NSCoder) {
-        hub = MediaEditorHub()
         super.init(coder: aDecoder)
     }
 
+    public override func viewDidLoad() {
+        super.viewDidLoad()
+
+        modalTransitionStyle = .crossDissolve
+        modalPresentationStyle = .fullScreen
+        navigationBar.isHidden = true
+    }
+
     public func edit(from viewController: UIViewController? = nil, onFinishEditing: @escaping (UIImage, [MediaEditorOperation]) -> (), onCancel: (() -> ())? = nil) {
         self.onFinishEditing = onFinishEditing
         self.onCancel = onCancel
         viewController?.present(self, animated: true)
     }
 
+    private func setup() {
+        hub.onCancel = cancel
+
+        setupForAsync()
+
+        presentIfSingleImageAndCapability()
+    }
+
+    private func setupForAsync() {
+        if let thumb = asyncImage?.thumb {
+            hub.show(image: thumb)
+        } else {
+            asyncImage?.thumbnail(finishedRetrievingThumbnail: thumbnailAvailable)
```

Comment:
Disclaimer: the following, though unlikely, is definitely possible.

This call starts an asynchronous download of the thumbnail.  The same is true for the `full()` call below.  Hence it's possible for the completion callbacks to be executed in a different order than you'd expect.

I believe the callbacks have no safeguards in place for triggering out of order.

Two simple possibilities to make this safer could be:

1. Not setting the thumbnail if the full image is already available, or
2. Queueing the downloads one after another (download thumbnail, and when done download full image).

I've no particular preference.

I realize this feels like a bit of an edge case, but since the extra work for a safeguard is not much, and since the bug would be annoying / hard to debug / hard to understand / noisy, I'd strongly suggest we try to prevent it.

Reply:
Great catch!

Fixed in 7b68415574545e0f64cfaeb8e75e8bd1f1a81089

Reply:
Great commit.  Marking as resolved.